### ISSUE TITLE
Rename providers to plugins

### ISSUE THREAD

**Issue by roleyfoley** [hamlet-io/engine#1480]
## Expected Behaviour
Providers aren't confused with cloud providers or other definitions of providers

## Current Behaviour
Currently we have the ideas of providers which are contain implementations of hamlet extensions, when they were initially implemented there was a one to one mapping between a cloud provider and the nominated provider ( aws, azure ) however with our shared provider, testing providers and documentation providers this is no longer the case and leads to some confusion when talking about deploying to a provider and implementing services in a provider. 

## Possible Solution
Rename the current implementation of providers to either packages or plugins. This aligns with broader industry terminology for extending the functionality of a system and better represents what our providers are capable of 

## Context
https://github.com/hamlet-io/engine/issues/1463#issuecomment-734514915

**Comment by roleyfoley** [comment]
@RossMurr4y @ml019 What do you think? With our expansion of our provider usage and the requirement for them to be defined by users so that they know how there outputs were generated I think this would make this clearer. 

I remember a few conversations where I've had to make the distinction between cloud provider and a hamlet provider, so I think its worth talking about

**Comment by ml019** [comment]
If we are thinking of having a package manager, then switching plugin to package makes sense.

This kind of means package isn't available to replace provider.

Agree that renaming provider makes sense to keep it clear of cloud providers, though the idea was not just to limit it to them, e.g. a git hosting provider.

**Comment by roleyfoley** [comment]
> Agree that renaming provider makes sense to keep it clear of cloud providers, though the idea was not just to limit it to them, e.g. a git hosting provider.

I think that would align with the point of renaming it, a plugin extends hamlet, that might be to add another cloud provider, it might be to add git hosting or it might be to draw documentation, its not in the implementation of a given plugin to decide what it will implement not that it has to implement a provider

**Comment by RossMurr4y** [comment]
I like both, though leaning more towards plugin for the reasons you've stated already. Plus, a "plugin" has always felt to me like it provided additional capabilities to a product, but it did so in a self-contained manner. A "package" to me sounds like it might do this, but it also might be a collection of smaller things that accomplish a goal that is "the sum of their parts" - i.e you can go to a package manager and download a package that is just a library, or you can download a package that is a full-featured application. A "plugin" is never the latter, whereas a "package" as we may later use it from a package-manager may be a single "module" or it may be many that together address a need of some sort.

TL;DR - agree with plugin
